fedimint-core/src/task/jit.rs

```diff
@@ -87,89 +86,42 @@ where
-                                trace!(target: LOG_TASK,  r#type = %type_name, "JitTry value ready, but ignored");
-                            },
-                        };
-                    },
-                }
-            },
+            async move { f().await },
         );
 
         Self {
-            _cancel_tx: cancel_tx,
-            val_rx: Arc::new(Some(val_rx).into()),
-            val: sync::OnceCell::new().into(),
+            inner: JitInner {
+                handle: handle.into(),
+                val: sync::OnceCell::new(),
+            }
+            .into(),
         }
     }
 
     /// Get the reference to the value, potentially blocking for the
     /// initialization future to complete
-    pub async fn get_try(&self) -> std::result::Result<&T, OneTimeError<E>> {
-        /// Temporarily taken data out of `arc`, that will be put back on drop,
-        /// unless canceled
-        ///
-        /// Used to achieve cancelation safety.
-        struct PutBack<'a, T> {
-            val: Option<T>,
-            arc: &'a Arc<std::sync::Mutex<Option<T>>>,
-        }
-
-        impl<'a, T> PutBack<'a, T> {
-            fn take(arc: &'a Arc<std::sync::Mutex<Option<T>>>) -> Self {
-                let val = arc.lock().expect("lock failed").take();
-                Self { val, arc }
-            }
-
-            fn cancel(mut self) {
-                self.val = None;
-            }
-
-            fn get_mut(&mut self) -> Option<&mut T> {
-                self.val.as_mut()
-            }
-        }
-        impl<'a, T> Drop for PutBack<'a, T> {
-            fn drop(&mut self) {
-                let mut lock = self.arc.lock().expect("lock failed");
-                let take = self.val.take();
-
-                *lock = take;
-            }
-        }
-        self.val
-            .get_or_try_init(|| async {
-                let mut recv = PutBack::take(&self.val_rx);
-
-                let val_res = {
-                    let Some(recv) = recv.get_mut() else {
-                        return Err(OneTimeError(None));
-                    };
-                    recv.await.unwrap_or_else(|_| {
-                        panic!("Jit value {} panicked", std::any::type_name::<T>())
+    pub async fn get_try(&self) -> Result<&T, OneTimeError<E>> {
+        let mut init_error = None;
+        let value = self
+            .inner
+            .val
+            .get_or_init(|| async {
+                let handle: &mut _ = &mut *self.inner.handle.lock().await;
+                handle
+                    .await
+                    .unwrap_or_else(|_| panic!("Jit value {} panicked", std::any::type_name::<T>()))
+                    .map_err(|err| {
+                        init_error = Some(err);
+                        // return ()
                     })
-                };
-
-                recv.cancel();
-
-                val_res.map_err(|err| OneTimeError(Some(err)))
             })
-            .await
+            .await;
```

Comment:
I had some doubts if *technically* it's cancleation safe and doesn't lose the error, but I think it's OK.

No idea why the you don't like the `Drop` thing so much. It is lenghtly, because it uses a lot of boilerplate. If it was a `scopeguard` it could have been a oneliner-ish, but didn't want to bring a new crate for that.